From: Jan Blunck <jblunck@suse.de>
Subject: Fix reference counting in __find_device_hash_cell()
References: 214682
Patch-mainline: no

When the device is created but it hasn't yet been inserted into the hash list
a concurrent call to __find_device_hash_cell() was failing to release the
reference to the mapped device.

Signed-off-by: Jan Blunck <jblunck@suse.de>
---
 drivers/md/dm-ioctl.c |    2 ++
 1 file changed, 2 insertions(+)

Index: linux-2.6.18/drivers/md/dm-ioctl.c
===================================================================
--- linux-2.6.18.orig/drivers/md/dm-ioctl.c
+++ linux-2.6.18/drivers/md/dm-ioctl.c
@@ -608,6 +608,8 @@ static struct hash_cell *__find_device_h
 	md = dm_get_md(huge_decode_dev(param->dev));
 	if (md)
 		mdptr = dm_get_mdptr(md);
+	if (md && !mdptr)
+		dm_put(md);
 
 	return mdptr;
 }
